07_errors/n0npax/store_test.go

```diff
@@ -0,0 +1,189 @@
+package main
+
+import (
+	"testing"
+
+	tassert "github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+var (
+	puppy0 = func() Puppy {
+		return Puppy{
+			Breed:  "Type A",
+			Colour: "Grey",
+			Value:  300,
+		}
+	}
+	puppy1 = func() Puppy {
+		return Puppy{
+			Breed:  "Type B",
+			Colour: "Brown",
+			Value:  400,
+		}
+	}
+	puppyNegativeValue = func() Puppy {
+		return Puppy{
+			Breed:  "Type C",
+			Colour: "Red",
+			Value:  -1,
+		}
+	}
+	puppyNegativeID = func() Puppy {
+		return Puppy{
+			Breed:  "Type D",
+			Colour: "Blue",
+			Value:  100,
+			ID:     -1,
+		}
+	}
+)
+
+type storerSuite struct {
+	suite.Suite
+	store Storer
+}
+
+func (s *storerSuite) SetupTest() {
+	switch s.store.(type) {
+	case *SyncStore:
+		s.store = NewSyncStore()
+	case *MemStore:
+		s.store = NewMemStore()
+	default:
+		panic("Unrecognised storer implementation")
+	}
+	puppy := puppy0()
+	_, err := s.store.CreatePuppy(&puppy)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func TestStorer(t *testing.T) {
+	suite.Run(t, &storerSuite{store: NewMemStore()})
+	suite.Run(t, &storerSuite{store: NewSyncStore()})
+}
+
+func (s *storerSuite) TestCreatePuppySuccessful() {
```

Comment:
You don't need to apply this but I thought I would share cause I think it is pretty cool. @anzboi and I recently came up with a format for structuring tests like this one which is kind of similar to the case based testing. The idea is to get the nicely structured test case which you get when creating subtests but to do it for a range of different use cases.

```go
func TestCreatePuppy(t *testing.T) {
  t.Run("Successful Creation", SuccessfulPuppyCreation)
  t.Run("Fails with Negative Value", CreateWithNegativeValue)
}

func (s *storerSuite) SuccessfulPuppyCreation(){
  assert := tassert.New(s.T())
  newPuppy0, newPuppy1 := puppy0(), puppy1()
  id0, err := s.store.CreatePuppy(&newPuppy0)
  assert.NoError(err, "Creating puppy should be ok")
  id1, err := s.store.CreatePuppy(&newPuppy1)
  assert.NoError(err, "Creating puppy should be ok")
  assert.Equal(id0, id1-1, "2nd id should be 1st +1, got %v and %v", id0, id1)
}

func (s *storerSuite) CreateWithNegativeValue() {
  assert := tassert.New(s.T())
  newPuppy := puppyNegativeValue()
  _, err := s.store.CreatePuppy(&newPuppy)
  assert.Error(err, "negative ID should cause an error")
}
```

You get an output which is formatted nicely, as well as each test being discrete and covering a single use case.

```bash
--- PASS: TestCreatePuppy (0.00s)
    --- PASS: TestCreatePuppy/Successful_Creation (0.00s)
    --- PASS: TestCreatePuppy/Fails_with_Negative_Value (0.00s)
```

Reply:
Thx. Will fix this in lab10